# Exercise 4.7 Deriving New Variables

This script contains the following points:
1. Create subset of df
2. If-statements with user-defined function
3. If-statements with loc() function
4. Missing values?
5. Task 4.7

In [5]:
#import libraries

import pandas as pd
import numpy as np
import os

## Import dataframe: merged_data

In [7]:
path = r'C:\Users\Barb\07-2024 Instacart Basket Analysis'

In [8]:
df_ords_prods_merge = pd.read_pickle(os.path.join(path, '02 Data', 'Prepared', 'ords_prods_merge.pkl'))

### Check output

In [10]:
df_ords_prods_merge.shape

(30328763, 13)

In [11]:
df_ords_prods_merge.head()

order_id  user_id  order_number  orders_day_of_week  hour_time_of_order  \
0   2398795        1             2                   3                   7   
1   2398795        1             2                   3                   7   
2   2398795        1             2                   3                   7   
3   2398795        1             2                   3                   7   
4   2398795        1             2                   3                   7   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                    15.0         196                  1          1   
1                    15.0       10258                  2          0   
2                    15.0       12427                  3          1   
3                    15.0       13176                  4          0   
4                    15.0       26088                  5          1   

                 product_name  aisle_id  department_id  prices  
0                        Soda        77              7     9.0  
1                  Pistachios       117             19     3.0  
2         Original Beef Jerky        23             19     4.4  
3      Bag of Organic Bananas        24              4    10.3  
4  Aged White Cheddar Popcorn        23             19     4.7

## 1. Create a subset of the dataframe

In [13]:
df = df_ords_prods_merge[:1000000]

In [14]:
df.shape

(1000000, 13)

In [15]:
df.head()

order_id  user_id  order_number  orders_day_of_week  hour_time_of_order  \
0   2398795        1             2                   3                   7   
1   2398795        1             2                   3                   7   
2   2398795        1             2                   3                   7   
3   2398795        1             2                   3                   7   
4   2398795        1             2                   3                   7   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                    15.0         196                  1          1   
1                    15.0       10258                  2          0   
2                    15.0       12427                  3          1   
3                    15.0       13176                  4          0   
4                    15.0       26088                  5          1   

                 product_name  aisle_id  department_id  prices  
0                        Soda        77              7     9.0  
1                  Pistachios       117             19     3.0  
2         Original Beef Jerky        23             19     4.4  
3      Bag of Organic Bananas        24              4    10.3  
4  Aged White Cheddar Popcorn        23             19     4.7

## 2. If-Statements with User-Defined Functions (with subset)

### 2.1 Define price label function to determine price range in subset

In [18]:
def price_label(row):

    if row['prices'] <= 5:
        return 'Low-range product'
    elif (row['prices'] > 5) and (row['prices'] <= 15):
        return 'Mid-range product'
    elif row['prices'] > 15:
        return 'High range product'
    else: return 'Not enough data'

### 2.2 Apply the price label function on subset

In [20]:
df['price_range'] = df.apply(price_label, axis=1)

C:\Users\Barb\AppData\Local\Temp\ipykernel_9036\2848743372.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['price_range'] = df.apply(price_label, axis=1)


### 2.3 Check values in new 'price_range' column in subset

In [22]:
df['price_range'].value_counts(dropna = False)

price_range
Mid-range product     672722
Low-range product     314975
High range product     12303
Name: count, dtype: int64

### 2.4 How much is the most expensive product within the subset?

In [24]:
df['prices'].max()

99999.0

## 3. If-Statements with the loc() Function (applied to subset)

In [26]:
df.loc[df['prices'] > 15, 'price_range_loc'] = 'High-range product'

C:\Users\Barb\AppData\Local\Temp\ipykernel_9036\1169838859.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df.loc[df['prices'] > 15, 'price_range_loc'] = 'High-range product'


In [27]:
df.loc[df['prices'] > 5 & (df['prices'] <= 15), 'price_range_loc'] = 'Mid-range product'

In [28]:
df.loc[df['prices'] <= 5, 'price_range_loc'] = 'Low-range product'

### 3.1 Check values in new 'price_range_loc' column of subset

In [30]:
df['price_range_loc'].value_counts(dropna = False)

price_range_loc
Mid-range product    685025
Low-range product    314975
Name: count, dtype: int64

## 4. Missing values?

### 4.1 Find missing values in entire dataframe df_ords_prods_merge

In [33]:
df_ords_prods_merge.isnull().sum()

order_id                  0
user_id                   0
order_number              0
orders_day_of_week        0
hour_time_of_order        0
days_since_prior_order    0
product_id                0
add_to_cart_order         0
reordered                 0
product_name              0
aisle_id                  0
department_id             0
prices                    0
dtype: int64

In [34]:
df_ords_prods_merge.shape

(30328763, 13)

# Task 4.7

### Task #1. Create the "price_label" and "busiest_day" columns

### 1.1 Create "price_label" column:     If-Statements with the loc() function (applied to entire 'ords_prods_merge' df)

In [38]:
df_ords_prods_merge.loc[df_ords_prods_merge['prices'] > 15, 'price_range_loc'] = 'High-range product'

In [39]:
df_ords_prods_merge.loc[df_ords_prods_merge['prices'] <= 15 & (df_ords_prods_merge['prices'] > 5), 'price_range_loc'] = 'Mid-range product'

In [40]:
df_ords_prods_merge.loc[df_ords_prods_merge['prices'] <= 5, 'price_range_loc'] = 'Low-range product'

### 1.1.1 Check values in new 'price_range_loc' column using frequency count

In [42]:
df_ords_prods_merge['price_range_loc'].value_counts(dropna = False)

price_range_loc
NaN                   20462144
Low-range product      9476774
High-range product      389845
Name: count, dtype: int64

### 1.2 Create 'busiest_day' column:  If-Statements with For_Loops

### 1.2.1 Find day most orders take place

In [45]:
df_ords_prods_merge['orders_day_of_week'].value_counts(dropna = False)

orders_day_of_week
0    5779087
1    5303718
6    4190948
5    3952326
2    3947564
3    3600589
4    3554531
Name: count, dtype: int64

### 1.2.2 Create an empty list - for the results. This acts as the empty shell in to which you can place your 'busiest_day' results.

In [47]:
result = []

In [48]:
for value in df_ords_prods_merge['orders_day_of_week']:
    if value == 0:
        result.append('Busiest day')
    elif value == 4:
        result.append('Least busy')
    else:
        result.append('Regularly busy')

In [49]:
result

['Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Reg

### 1.2.3 Combine 'busiest_day' column with 'ords_prods_merge' dataframe

In [51]:
df_ords_prods_merge['busiest_day'] = result

### 1.2.4 Check values in new 'busiest day' column

In [53]:
df_ords_prods_merge['busiest_day'].value_counts(dropna = False)

busiest_day
Regularly busy    20995145
Busiest day        5779087
Least busy         3554531
Name: count, dtype: int64

### 1.2.5. Check output

In [55]:
df_ords_prods_merge.head()

order_id  user_id  order_number  orders_day_of_week  hour_time_of_order  \
0   2398795        1             2                   3                   7   
1   2398795        1             2                   3                   7   
2   2398795        1             2                   3                   7   
3   2398795        1             2                   3                   7   
4   2398795        1             2                   3                   7   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                    15.0         196                  1          1   
1                    15.0       10258                  2          0   
2                    15.0       12427                  3          1   
3                    15.0       13176                  4          0   
4                    15.0       26088                  5          1   

                 product_name  aisle_id  department_id  prices  \
0                        Soda        77              7     9.0   
1                  Pistachios       117             19     3.0   
2         Original Beef Jerky        23             19     4.4   
3      Bag of Organic Bananas        24              4    10.3   
4  Aged White Cheddar Popcorn        23             19     4.7   

     price_range_loc     busiest_day  
0                NaN  Regularly busy  
1  Low-range product  Regularly busy  
2  Low-range product  Regularly busy  
3                NaN  Regularly busy  
4  Low-range product  Regularly busy

### Task #2 Suppose your clients have changed their minds about the labels you created in your “busiest_day” column. Now, they want “Busiest day” to become “Busiest days” (plural). This label should correspond with the two busiest days of the week as opposed to the single busiest day. At the same time, they’d also like to know the two slowest days. Create a new column for this using a suitable method.

### 2.1 Find day most orders take place

In [58]:
df_ords_prods_merge['orders_day_of_week'].value_counts(dropna = False)

orders_day_of_week
0    5779087
1    5303718
6    4190948
5    3952326
2    3947564
3    3600589
4    3554531
Name: count, dtype: int64

### 2.2 Create an empty list - for the results. This acts as the empty shell in to which you can place your 'busiest_days' results.¶

In [60]:
result = []

In [61]:
for value in df_ords_prods_merge['orders_day_of_week']:
    if value == 0 or value == 1:
        result.append('Busiest days')
    elif value == 3 or value == 4:
        result.append('Slowest days')
    else:
        result.append('Regularly busy')

### 2.3 Combine column of 'busiest_days' results with 'ords_prods_merge' dataframe

In [63]:
df_ords_prods_merge['busiest_days'] = result

### Task #3 Check values of this new 'busiest days' column for accuracy. Note any observations in markdown format.

In [65]:
df_ords_prods_merge['busiest_days'].value_counts(dropna = False)

busiest_days
Regularly busy    12090838
Busiest days      11082805
Slowest days       7155120
Name: count, dtype: int64

### Total days according to frequency count above = 30,328,763
### Total rows in dataframe 'ords_prods_merge' = 30,328,763

### Task #4 - When too many users make Instacart orders at the same time, the app freezes. The senior technical officer at Instacart wants you to
### identify the busiest hours of the day. Rather than by hour, they want periods of time labeled "most orders," "average orders," and "fewest orders." ### Create a new column containing these labels called "busiest_period_of_day."

### 4.1 Find periods of day where most orders take place

In [69]:
df_ords_prods_merge['hour_time_of_order'].value_counts(dropna = False)

hour_time_of_order
10    2593725
11    2564597
14    2517238
15    2487586
13    2487500
12    2445841
16    2364969
9     2311334
17    1943858
8     1622394
18    1520954
19    1169224
20     910005
7      844665
21     746254
22     592432
23     375889
6      274801
0      203460
1      108110
5       82706
2       63961
4       49400
3       47860
Name: count, dtype: int64

### 4.2 Create an empty list - for the results. This acts as the empty shell in to which you can place your 'busiest_period_of_day' results.

In [71]:
result = []

In [72]:
for value in df_ords_prods_merge['hour_time_of_order']:
    if value == 10 or value == 11 or value == 14 or value == 15 or value == 13 or value == 12:
        result.append('Most Orders')
    elif value == 3 or value == 4 or value == 2 or value == 5 or value == 1 or value == 0 or value == 6 or value == 23:
        result.append('Fewest Orders')
    else:
        result.append('Average Orders')

### 4.3 Combine column of 'busiest_period_of_day' results with 'ords_prods_merge' dataframe

In [74]:
df_ords_prods_merge['busiest_period_of_day'] = result

### 4.4 Check output

In [76]:
df_ords_prods_merge.head()

order_id  user_id  order_number  orders_day_of_week  hour_time_of_order  \
0   2398795        1             2                   3                   7   
1   2398795        1             2                   3                   7   
2   2398795        1             2                   3                   7   
3   2398795        1             2                   3                   7   
4   2398795        1             2                   3                   7   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                    15.0         196                  1          1   
1                    15.0       10258                  2          0   
2                    15.0       12427                  3          1   
3                    15.0       13176                  4          0   
4                    15.0       26088                  5          1   

                 product_name  aisle_id  department_id  prices  \
0                        Soda        77              7     9.0   
1                  Pistachios       117             19     3.0   
2         Original Beef Jerky        23             19     4.4   
3      Bag of Organic Bananas        24              4    10.3   
4  Aged White Cheddar Popcorn        23             19     4.7   

     price_range_loc     busiest_day  busiest_days busiest_period_of_day  
0                NaN  Regularly busy  Slowest days        Average Orders  
1  Low-range product  Regularly busy  Slowest days        Average Orders  
2  Low-range product  Regularly busy  Slowest days        Average Orders  
3                NaN  Regularly busy  Slowest days        Average Orders  
4  Low-range product  Regularly busy  Slowest days        Average Orders

### Task #5. Print frequency for this new column

In [78]:
df_ords_prods_merge['busiest_period_of_day'].value_counts(dropna = False)

busiest_period_of_day
Most Orders       15096487
Average Orders    14026089
Fewest Orders      1206187
Name: count, dtype: int64

### Task #7. Export your database as a pickle file and store it in your "Prepared Data" folder

In [80]:
df_ords_prods_merge.to_pickle(os.path.join(path, '02 Data','Prepared', 'ords_prods_merge_4.7.pkl'))